tests/test_reference_velocity_verlet_integrator.py

```diff
@@ -0,0 +1,129 @@
+    """Assert reversibility of .step and .multiple_steps implementations"""
+
+    # check step implementation
+    def step_update(x, v):
+        for t in range(n_steps):
+            x, v = intg.step(x, v)
+        return x, v
+
+    assert_reversible(x0, v0, step_update, atol=atol)
+
+    # check multiple_steps implementation
+    def multiple_steps_update(x, v):
+        xs, vs = intg.multiple_steps(x, v, n_steps)
+        return xs[-1], vs[-1]
+
+    assert_reversible(x0, v0, multiple_steps_update, atol=atol)
+
+
+def test_reversibility_with_jax_potentials():
+    """On a simple jax-transformable potential (quartic oscillators)
+    with randomized parameters and initial conditions
+    (n oscillators, masses, dt, x0, v0)
+    assert all 3 update functions
+    (public: .step and .multiple_steps, private: ._update_via_fori_loop)
+    are reversible"""
+
+    np.random.seed(2022)
+
+    def U(x):
+        return jnp.sum(x ** 4)
+
+    @jit
+    def force(x):
+        return -grad(U)(x)
+
+    for n_steps in [1, 10, 100, 1000, 10000]:
+        n = np.random.randint(10, 10000)  # Unif[10, 10000]
+        masses = np.random.rand(n) + 1  # Unif[1, 2]
+        dt = 0.09 * np.random.rand() + 0.01  # Unif[0.01, 0.1]
+        x0 = np.random.randn(n, 3)
+        v0 = np.random.randn(n, 3)
+
+        intg = VelocityVerletIntegrator(force, masses, dt)
+
+        # check "public" .step and .multiple_steps implementations
+        assert_reversibility_using_step_implementations(intg, x0, v0, n_steps)
+
+        # also check "private" jax.lax.fori_loop implementation
+        @jit
+        def jax_update(x, v):
+            return intg._update_via_fori_loop(x, v, n_steps)
+
+        assert_reversible(x0, v0, jax_update, atol=1e-10)
+
+
+def test_reversibility_with_custom_ops_potentials():
+    """Check reversibility of "public" .step and .multiple_steps implementations when `force_fxn`
+    is a custom_op potential"""
+
+    np.random.seed(2022)
+
+    # define a Python force fxn that calls custom_ops
+    rfe = hif2a_ligand_pair
+    unbound_potentials, sys_params, masses = rfe.prepare_vacuum_edge(rfe.ff.get_ordered_params())
+    coords = rfe.prepare_combined_coords()
+    bound_potentials = [
+        ubp.bind(params).bound_impl(np.float32) for (ubp, params) in zip(unbound_potentials, sys_params)
+    ]
+    box = 100 * np.eye(3)
+
+    def force(coords):
+        du_dxs = np.array([bp.execute(coords, box, 0.5)[0] for bp in bound_potentials])
+        return -np.sum(du_dxs, 0)
+
+    dt = 1.5e-3
+    intg = VelocityVerletIntegrator(force, masses, dt)
+
+    x0 = np.array(coords)
+
+    for n_steps in [1, 10, 100, 500, 1000]:  # , 10000]:
+        v0 = np.random.randn(*coords.shape)
+
+        # check "public" .step and .multiple_steps implementations
+        assert_reversibility_using_step_implementations(intg, x0, v0, n_steps, atol=1e-10)
+
+    # TODO: possibly investigate why n_steps = 10000 fails
```

Comment:
after some offline discussion with @maxentile  - we can definitely make this bitwise reversible if we do accumulation in fixed point.